_posts/2019-04-16-accessing-kafka-part-1.md

```diff
@@ -0,0 +1,107 @@
+
+So how does Strimzi deal with this?
+There are two general ways how to deal with this problem:
+* Write your own proxy / load balancer which would do more intelligent routing on application layer (layer 7). Such proxy can for example abstract the architecture of the Kafka cluster from the client and pretend that the cluster has just one big broker running everything and just route the traffic to the different brokers in the background. Kubernetes already does this for the HTTP traffic using the Ingress resource.
+* Make sure you use the `advertised.listeners` option in the broker configuration in a way which allows the clients to connect directly to the broker.
+
+In Strimzi, we currently support the second option.
+
+## Connecting from inside the same Kubernetes cluster
+
+Doing this for the clients running inside the same Kubernetes cluster as the Kafka cluster is running is actually quite simple.
+Each pod has its own IP address, which other applications can use to connect directly to it. 
+This is normally not used by regular Kubernetes applications.
+One of the reasons is that Kubernetes doesn't offer any nice way how to discover these IP addresses.
+To find out the IP address, you would need to use the Kubernetes API, find the right pod and its IP address.
+And you would need to have the rights for this.
+Instead, Kubernetes use the services with their stable DNS names as the main discovery mechanism.
+But with Kafka, this not an issue, because it has its own discovery protocol.
+We do not need the clients to figure out the API address from the Kubernetes API.
+We just need to configure it ad the advertised address and the clients will discover it through the Kafka _metadata_.
+
+But there is even one better option which is used by Strimzi.
+For StatefulSet - which Strimzi is using to run the Kafka brokers - you can use the Kubernetes headless service to give each of the pods a stable DNS name.
+Strimzi is using these DNS names as the advertised addresses for the Kafka brokers.
+So with Strimzi:
+* The initial connection is done using a regular Kubernetes service to get the _metadata_.
+* The subsequent connections are opened using the DNS names give to the pods by another headless Kubernetes service.
+The diagram below shows how does it look with an example Kafka cluster named `my-cluster`.
+
+![Accessing Kafka inside the same Kubernetes cluster]({{ "/assets/2019-04-16-inside-kubernetes.png" }})
+
+Both approaches have their own pros and cons.
+Using the DNS can sometimes cause problems with cached DNS information.
+When the underlying IP addresses of the pods change for example during rolling updates, the clients connecting to the brokers need to have the latest DNS information.
+However, we found out the using the IP addresses is causing even worse problems, because sometimes Kubernetes re-use them very aggressively and a new pod gets the IP address used just few seconds before by some other Kafka node.
```

Comment:
```suggestion
However, we found out that using IP addresses is causes even worse problems, because sometimes Kubernetes re-uses them very aggressively and a new pod gets the IP address used just a few seconds before by some other Kafka node.
```

Reply:
I guess we should also remove the `is` in `is causes`, right?